### PR TITLE
Adds main abstract classes related to HSMs

### PR DESCRIPTION
**Pull request by johnfxgalea-e**
Adds two main abstract classes concerning HSMs. The `HSM` class is the highest abstract class in the hierarchy; all HSM implementations should inherit it either directly or indirectly. The second abstract class is the `HsmEngine` class which extends `HSM` and adds functionality based on OpenSSL's `ENGINE_* API`.

Since these two classes are both abstract, no concrete implementation of an HSM is provided in this PR. Such an implementation will be provided in a subsequent PR to ease review.

### CODE REVIEW

File: src/mococrw/hsm.h [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+    /**
+     * Generates a key pair via HSM.
+     *
+     * \note Currently, RSA keys are only supported.
+     */
+    virtual void generateKey(unsigned int bits,
+                             const std::string &label,
+                             const std::string &id) = 0;
+
+    // Many of protected functions provided by the HSM class are seen
```
Comment (wusto) [inline src/mococrw/hsm.h]: I guess the friend class belong logically to a different PR.

Please add them where the friend classes access the protected functions.

File: src/mococrw/hsm.h [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+    virtual ~HSM() {}
+
+protected:
+    /**
+     * Returns the name of the HSM driver.
+     */
+    virtual const std::string getName() = 0;
+
+    /**
+     *  Loads public key from HSM.
```
Comment (wusto) [inline src/mococrw/hsm.h:46]: Wouldn't it make sense to add descriptions for the parameters or at least a link to the official documentation, which describes the mechanisms behind the arguments?

File: src/mococrw/hsm.h [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+    /** Pointer to OpenSSL ENGINE. */
+    openssl::SSL_ENGINE_Ptr _engine;
+    /** Engine ID. */
+    const std::string &_id;
+    /** Path to Module. */
+    const std::string &_modulePath;
+    /** Pin to access PKCS11 Engine. */
+    const std::string &_pin;
+
+    virtual const std::string getName() override = 0;
```
Comment (wusto) [inline src/mococrw/hsm.h]: So this class is still an abstract one?

From the class is description it reads like this is the actual implementation.

File: src/mococrw/hsm.h [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+protected:
+    /**
+     * Returns the name of the HSM driver.
+     */
+    virtual const std::string getName() = 0;
+
+    /**
+     *  Loads public key from HSM.
+     */
+    virtual openssl::SSL_EVP_PKEY_Ptr loadPublicKey(const std::string &keyID) = 0;
```
Comment (wusto) [inline src/mococrw/hsm.h:50]: What's the rational to return a openssl::SSL_EVP_PKEY_Ptr and not an object of type AsymmetricPublicKey or AsymmetricKeypair?

Usually this requires then an additional layer of abstraction to cast/transform it into the key object(s) used by MoCOCrW.

**Review by wusto — CHANGES_REQUESTED** [review]

File: src/mococrw/hsm.h (hunk 1 above)
Comment (johnfxgalea-e) [reply src/mococrw/hsm.h]: Makes sense. I will remove.

File: src/mococrw/hsm.h (hunk 4 above)
Comment (johnfxgalea-e) [reply src/mococrw/hsm.h:50]: This is an internal API which is to be used by AsymmetricPublicKey/AsymmetricKeypair (this is where the friend relationship comes in and why these functions are marked protected).

File: src/mococrw/hsm.h (hunk 1 above)
Comment (johnfxgalea-e) [reply src/mococrw/hsm.h]: Done.

File: src/mococrw/hsm.h (hunk 2 above)
Comment (johnfxgalea-e) [reply src/mococrw/hsm.h:46]: Done.

File: src/mococrw/hsm.h (hunk 3 above)
Comment (johnfxgalea-e) [reply src/mococrw/hsm.h]: Done.

File: src/mococrw/hsm.h [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ * #L%
+ */
+#pragma once
+
+#include <boost/optional/optional.hpp>
```
Comment (istepic) [inline src/mococrw/hsm.h]: Where is this used?

File: src/hsm.cpp [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ * #L%
+ */
+#include "mococrw/hsm.h"
+#include <boost/format.hpp>
```
Comment (istepic) [inline src/hsm.cpp]: Where is this used?

File: src/mococrw/hsm.h [hunk 7]
```diff
@@ -0,0 +1,120 @@
...
+
+namespace mococrw
+{
+/**
+ * Driver for Hardware Security Modules (HSMs).
+ */
+class HSM
+{
+public:
+    HSM();
```
Comment (istepic) [inline src/mococrw/hsm.h]: Not needed.

File: src/hsm.cpp [hunk 8]
```diff
@@ -0,0 +1,51 @@
...
+
+namespace mococrw
+{
+using namespace openssl;
+
+HSM::HSM() {}
```
Comment (istepic) [inline src/hsm.cpp]: Not needed.

File: src/mococrw/hsm.h [hunk 9]
```diff
@@ -0,0 +1,120 @@
...
+ */
+#pragma once
+
+#include <boost/optional/optional.hpp>
+#include "openssl_wrap.h"
+
+namespace mococrw
+{
+/**
+ * Driver for Hardware Security Modules (HSMs).
```
Comment (istepic) [inline src/mococrw/hsm.h]: Can you update/be more verbose in this comment? This class has no implementation so I wouldn't say it's a driver for anything.

File: src/mococrw/hsm.h [hunk 10]
```diff
@@ -0,0 +1,120 @@
...
+ * HSM driver that leverages OpenSSL's ENGINE_* API interface.
+ */
+class HsmEngine : public HSM
+{
+public:
+    HsmEngine(const std::string &id, const std::string &modulePath, const std::string &pin);
+    virtual ~HsmEngine();
+
+    ENGINE *internal();
+    const ENGINE *internal() const;
```
Comment (istepic) [inline src/mococrw/hsm.h]: Please rethink why would we want to give ptr to ENGINE to clients. This is currently not used anywhere and the association between `ENGINE*` and `openssl::SSL_ENGINE_Ptr` is not clear to me.

File: src/mococrw/hsm.h [hunk 11]
```diff
@@ -0,0 +1,120 @@
...
+    /** Pointer to OpenSSL ENGINE. */
+    openssl::SSL_ENGINE_Ptr _engine;
+    /** Engine ID. */
+    const std::string &_id;
+    /** Path to Module. */
+    const std::string &_modulePath;
+    /** Pin to access PKCS11 Engine. */
+    const std::string &_pin;
+
+    virtual const std::string getName() override = 0;
```
Comment (istepic) [inline src/mococrw/hsm.h]: It's tough to see use-case for this function and why it's not implemented.

**Review by garw — APPROVED** [review]
lg2m

File: src/mococrw/hsm.h [hunk 12]
```diff
@@ -0,0 +1,120 @@
...
+
+    virtual openssl::SSL_EVP_PKEY_Ptr loadPublicKey(const std::string &keyID) override;
```
Comment (istepic) [inline src/mococrw/hsm.h]: [Avoid using virtual and override.](https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#c128-virtual-functions-should-specify-exactly-one-of-virtual-override-or-final)

File: src/mococrw/hsm.h [hunk 13]
```diff
@@ -0,0 +1,120 @@
...
+
+    virtual openssl::SSL_EVP_PKEY_Ptr loadPrivateKey(const std::string &keyID) override;
```
Comment (istepic) [inline src/mococrw/hsm.h]: s.a.

File: src/mococrw/hsm.h [hunk 14]
```diff
@@ -0,0 +1,120 @@
...
+namespace mococrw
+{
+/**
+ * Driver for Hardware Security Modules (HSMs).
+ */
+class HSM
+{
+public:
+    HSM();
+    virtual ~HSM() {}
```
Comment (istepic) [inline src/mococrw/hsm.h]: nitpick: Usually `= default;` is used in this codebase.

File: src/mococrw/hsm.h [hunk 15]
```diff
@@ -0,0 +1,120 @@
...
+    virtual openssl::SSL_EVP_PKEY_Ptr loadPrivateKey(const std::string &keyID) = 0;
+
+    /**
+     * Stores private key to HSM.
+     *
+     * @param key The private key to store.
+     * @param label The label of the key to store.
+     * @param keyID The ID of the key to store.
+     */
+    virtual void storePrivateKey(EVP_PKEY *key,
```
Comment (istepic) [inline src/mococrw/hsm.h:77]: Why is `EVP_PKEY*` used here and `openssl::SSL_EVP_PKEY_Ptr` in the function above? Preferably we would use only one for consistency unless this is how it's supposed to work.

File: src/hsm.cpp [hunk 16]
```diff
@@ -0,0 +1,51 @@
...
+{
+    // Fetch engine via ID.
+    _engine = _ENGINE_by_id(_id);
+
+    _ENGINE_ctrl_cmd_string(_engine.get(), "MODULE_PATH", _modulePath);
+    _ENGINE_ctrl_cmd_string(_engine.get(), "PIN", _pin);
+    _ENGINE_init(_engine.get());
+}
+
+HsmEngine::~HsmEngine() { _ENGINE_finish(_engine.get()); }
```
Comment (istepic) [inline src/hsm.cpp:41]: What about _ENGINE_free()?

**Review by istepic — DISMISSED** [review]
Looks good overall but I would like to see a slightly expanded version of this PR which tests at least one of the HSMEngine methods. You did provide implementation for HSMEngine so this class is not really abstract and it would be good to see if construction/destruction of HSMEngine works as intended. An example wouldn't hurt but that can be done later.

File: src/mococrw/hsm.h (hunk 3 above)
Comment (wusto) [reply src/mococrw/hsm.h]: Can you please recheck. Seems like a rebase issue.

**Review by wusto — CHANGES_REQUESTED** [review]

File: src/mococrw/hsm.h (hunk 9 above)
Comment (johnfxgalea-e) [reply src/mococrw/hsm.h]: Done.

File: src/mococrw/hsm.h (hunk 7 above)
Comment (johnfxgalea-e) [reply src/mococrw/hsm.h]: Done.

File: src/mococrw/hsm.h (hunk 14 above)
Comment (johnfxgalea-e) [reply src/mococrw/hsm.h]: Done.

File: src/mococrw/hsm.h (hunk 10 above)
Comment (johnfxgalea-e) [reply src/mococrw/hsm.h]: That's fair. I based it on a pattern of other existing classes in the code-base (e.g., `BioObject`), but I agree I don't need this functionality for now... Removing.

File: src/mococrw/hsm.h (hunk 11 above)
Comment (johnfxgalea-e) [reply src/mococrw/hsm.h]: I will implement this function by having the string be passed as a parameter via the construction.

File: src/mococrw/hsm.h (hunk 12 above)
Comment (johnfxgalea-e) [reply src/mococrw/hsm.h]: Done.

File: src/mococrw/hsm.h (hunk 13 above)
Comment (johnfxgalea-e) [reply src/mococrw/hsm.h]: Done.

File: src/mococrw/hsm.h (hunk 10 above)
Comment (johnfxgalea-e) [reply src/mococrw/hsm.h]: Done.

File: src/hsm.cpp (hunk 16 above)
Comment (johnfxgalea-e) [reply src/hsm.cpp:41]: That function is triggered via the smart pointer automatically

File: src/hsm.cpp (hunk 8 above)
Comment (johnfxgalea-e) [reply src/hsm.cpp]: N/A

File: src/hsm.cpp (hunk 6 above)
Comment (johnfxgalea-e) [reply src/hsm.cpp]: Done

File: src/mococrw/hsm.h (hunk 5 above)
Comment (johnfxgalea-e) [reply src/mococrw/hsm.h]: Removed. Done

File: src/mococrw/hsm.h (hunk 11 above)
Comment (johnfxgalea-e) [reply src/mococrw/hsm.h]: Done.

File: src/mococrw/hsm.h (hunk 3 above)
Comment (johnfxgalea-e) [reply src/mococrw/hsm.h]: I tried to make the code comment clearer to denote that this class is also abstract.

File: src/mococrw/hsm.h (hunk 15 above)
Comment (johnfxgalea-e) [reply src/mococrw/hsm.h:77]: This is the pattern that the rest of the code base follows for intermediary layers such as the one in this PR. For example, `keygen()` takes the raw OpenSSL pointer as parameter, but returns the wrapped version of EVP_PKEY:

`SSL_EVP_PKEY_Ptr _EVP_PKEY_keygen(EVP_PKEY_CTX *ctx)`

See: https://github.com/bmwcarit/MoCOCrW/blob/361f880ae068a0a346e9bb9301e0025ff97e2892/src/key.cpp#L120

**Review by wusto — APPROVED** [review]

**Comment by johnfxgalea-e** [comment]
> Looks good overall but I would like to see a slightly expanded version of this PR which tests at least one of the HSMEngine methods. You did provide implementation for HSMEngine so this class is not really abstract...

Not sure I follow. HSMEngine is not a concrete class, and indeed abstract.

File: src/hsm.cpp [hunk 17]
```diff
@@ -0,0 +1,53 @@
...
+ * limitations under the License.
+ * #L%
+ */
+#include "mococrw/hsm.h"
+
+namespace mococrw
+{
+using namespace openssl;
+
+const std::string &HSM::getName() { return _name; }
```
Comment (istepic) [inline src/hsm.cpp:25]: I still don't see a use-case for this _name attribute but I guess we'll see that in the next PRs.

**Review by istepic — COMMENTED** [review]
> > Looks good overall but I would like to see a slightly expanded version of this PR which tests at least one of the HSMEngine methods. You did provide implementation for HSMEngine so this class is not really abstract...
> 
> Not sure I follow. HSMEngine is not a concrete class, and indeed abstract.

Depends on the [definition](https://en.cppreference.com/w/cpp/language/abstract_class) you use I guess. Typically it's nice to see some tests with implementation in the same PR but since I know you have something in your backpocket I'm going to remain neutral here.

File: src/hsm.cpp (hunk 17 above)
Comment (johnfxgalea-e) [reply src/hsm.cpp:25]: This is not a vital function and not to be given significant importance. It is more with the intention of having a descriptive name, as a string, associated with concrete HSM classes (e.g., for logging).
